File: CivitasRuby/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require_relative 'enumerados'
require_relative 'respuestas'
require_relative 'salidas_carcel'
require_relative 'diario'
require 'io/console'

module Civitas

  class Vista_textual
    
    attr_reader :modelo
    
    def initialize
      @iGestion = -1
      @iPropiedad = -1
    end

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
                        
      return opcion
    end

    
    def comprar
      lista_Respuestas = [Respuestas::SI, Respuestas::NO]
      opcion = menu("¿Comprar calle?", ["SI", "NO"])
      return (lista_Respuestas[opcion])
      
    end

    def gestionar
    lista_gestiones_inmobiliarias = [OperacionesInmobiliarias::VENDER , 
    OperacionesInmobiliarias::HIPOTECAR, 
    OperacionesInmobiliarias::CANCELAR_HIPOTECA, 
    OperacionesInmobiliarias::CONSTRUIR_CASA, 
    OperacionesInmobiliarias::CONSTRUIR_HOTEL,
    OperacionesInmobiliarias::TERMINAR ]

      opcion = menu("¿Qué gestión inmobiliaria quiere realizar?", 
        ["Vender", "Hipotecar","Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar"])
      

      @iGestion = lista_gestiones_inmobiliarias[opcion]
      
      nombres_propiedades = Array.new
      
      for i in (0..@modelo.get_jugador_actual.propiedades.length - 1)
        p = @modelo.get_jugador_actual.propiedades
        puts p[i].nombre
        nombres_propiedades << @modelo.get_jugador_actual.propiedades.at(i).nombre
      end
      
      @iPropiedad = menu("¿En qué propiedad?" , nombres_propiedades)
    end

    def getGestion
      @iGestion
    end

    def getPropiedad
      @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
    
      case operacion
      when  OperacionesJuego::AVANZAR
         puts "---La siguiente operación es avanzar---"
      when  OperacionesJuego::COMPRAR
         puts "---La siguiente operación es comprar---"
      when OperacionesJuego::GESTIONAR
         puts "---La siguiente operación es gestionar---"
      when  OperacionesJuego::PASAR_TURNO
         puts "---La siguiente operación es pasar turno---"
      else
         puts "---La siguiente operación es salir de la cárcel---"
        
      end
    end

    def mostrarEventos
      
      
      while(Diario.instance.eventos_pendientes)
        puts Diario.instance.leer_evento
      end
      
    end

    def setCivitasJuego(civitas)
         @modelo=civitas
         actualizarVista
    end

    def actualizarVista
      system("clear")
      puts "***INFO_JUGADOR***"
      puts @modelo.get_jugador_actual.to_string
      puts "***INFO_CASILLA***"
      puts @modelo.get_casilla_actual.to_string
    end

    
    def salirCarcel
      lista_Salidas_carcel = [Salidas_carcel::PAGANDO, Salidas_carcel::TIRANDO]
      opcion = menu("Elige la forma para intentar salir de la cárcel", ["Pagando", "Tirando el dado"])
      return (lista_Salidas_carcel[opcion])
    end
    
  end

end

File: CivitasJava/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.Respuestas;
import civitas.GestionesInmobiliarias;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.CivitasJuego;
import civitas.Diario;
import civitas.Jugador;
import civitas.OperacionesJuego;
import civitas.Respuestas;
import civitas.SalidasCarcel;
import civitas.TituloPropiedad;

public class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  public VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado() {
    System.out.println (juegoModel.getEstado());
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() 
  {
      String titulo="¿Quiere comprar la calle?";
      ArrayList<String> lista=new ArrayList<String>();
      lista.add("SI");
      lista.add("NO");
      if(menu(titulo,lista)==0)//Si la opcion marcada es SI
      {
          return Respuestas.SI;
      }
      else
      {
          return Respuestas.NO;
      }
              
  }

  
  
  void gestionar () 
  {
      String titulo="\nGestión";
      ArrayList<String> lista;
      lista=new ArrayList<>();
      ArrayList<String> listaprop;
      listaprop = new ArrayList<>();
      lista.add("Vender");
      lista.add("Hipotecar");
      lista.add("Cancelar Hipoteca");      
      lista.add("Construir Casa");
      lista.add("Construir Hotel");
      lista.add("Terminar");   
      for(int i=0;i<juegoModel.getJugadorActual().getSizePropiedades();i++)
      {
          listaprop.add(juegoModel.getJugadorActual().getPropiedades().get(i).getNombre());
      }
      
      iGestion=menu(titulo,lista);
      if(!juegoModel.getJugadorActual().getPropiedades().isEmpty() && iGestion!=lista.indexOf("Terminar"))
      {
          iPropiedad=menu("Propiedad:",listaprop);
      }
      else if(iGestion!=lista.indexOf("Terminar"))
      {
          System.out.print("\nNo posees ninguna propiedad\n");
      }
         
  }
  
  
  
  public int getGestion()
  {
      return iGestion;
  }
  
  
  
  public int getPropiedad()
  {
      return iPropiedad;
  }
    

  
  void mostrarSiguienteOperacion(OperacionesJuego operacion) 
  {
      System.out.print(operacion);
  }
  
  
  
  void mostrarEventos()
  {
      if(Diario.getInstance().eventosPendientes())
      {
          System.out.print("\nDiario:");
          System.out.print(Diario.getInstance().leerEvento());
          System.out.print("\n");
      }
  }
 
  
  
  public void setCivitasJuego(CivitasJuego civitas)
    {
        
        juegoModel=civitas;
        actualizarVista();

    }
  
  
  
  
public void actualizarVista()
  {
      System.out.print("\n----ESTADO JUGADOR----\n");
      System.out.print(juegoModel.getJugadorActual().toString());
      System.out.print("\n----ESTADO CASILLA----\n");
      System.out.print(juegoModel.getCasillaActual().toString());
      System.out.print("\n----------------------\n");      
  } 
}
